### PR TITLE
minimal-bootstrap: use recursive FOD to make nix unpack bootstrap sources

### PR DESCRIPTION
**Pull request by emilytrau**
###### Description of changes

This is an addendum to work done in #238357. Investigating it seems like the contents of the narball wasn't being unpacked, causing hex0 to not find sources for hex1 and hang. Switching to a recursive FOD seems like a solution?

Related #227914

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix [hunk 1]
```diff
@@ -85,22 +85,13 @@ rec {
...
-      # You can create this file from an already-bootstrapped nixpkgs
+      # You can create this path from an already-bootstrapped nixpkgs
       # using the following command:
       #
       #   nix-build '<nixpkgs>' -A make-minimal-bootstrap-sources
       #
-      # Or, if you prefer, you can create this file using only `git`,
-      # `nix`, and `xz`.  For the commands needed in order to do this,
-      # see `make-bootstrap-sources.nix`.
-      #
```
Comment (amjoseph-nixpkgs) [inline pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix:103]: Being able to give the user these instructions is kinda important.  Let me see if I can find a way to get back to that point.

Comment (Ericson2314) [reply pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix:103]: We can have the same instructions for doing it manually with the other Nix store add command, no?

Comment (amjoseph-nixpkgs) [reply pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix:103]: That's what I'm going to try to figure out.

Comment (Ericson2314) [reply pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix]: I added them bakc.

**Review by Ericson2314 — APPROVED** [review]

**Comment by Ericson2314** [comment]
It builds again, but unfortunately I get a segfault building glibc.

**Comment by zimbatm** [comment]
I'm not sure we fully understand the implications of this change. It's effectively changing the position of cache.nixos.org from being a cache to being a source for bootstrapping nixpkgs.

The advantage is pretty obvious, but I think there is a downside where bootstrapping becomes harder if cache.nixos.org were to disappear or lost those files.

**Comment by Ericson2314** [comment]
I would say it's probably better to worry about such things when we're ready to switch over to this.

tarballs.nixos.org can be a substituter for fixed output things with a little bit of work, native git hashing support should hopefully make fetching these sources easier, etc.

**Comment by zimbatm** [comment]
Isn't `nix-build '<nixpkgs>' -A make-minimal-bootstrap-sources` relying on the previous tarball to build itself?

**Comment by amjoseph-nixpkgs** [comment]
> Isn't `nix-build '<nixpkgs>' -A make-minimal-bootstrap-sources` relying on the previous tarball to build itself?

Keep reading....

https://github.com/emilytrau/nixpkgs/blob/4179372632f8f5313ac8684690a0f636295e9146/pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix#L95-L97

Of course, it would be nice if we actually *printed* those lines, so the user could copy-paste them into the terminal.

**Pull request merged**